cfgov/login/auth.py

```diff
@@ -0,0 +1,158 @@
+
+    if settings.SSO_DEFAULT_GROUP is None:
+        logger.debug("SSO_DEFAULT_GROUP not set, not modifying user groups")
+        return
+
+    # All SSO users will belong to a group that can log into the admin
+    group = Group.objects.get_or_create(name=settings.SSO_DEFAULT_GROUP)[0]
+
+    # Make sure this group has permission to log into the admin
+    permission = Permission.objects.get(
+        codename="access_admin",
+        content_type=ContentType.objects.get(
+            app_label="wagtailadmin", model="admin"
+        ),
+    )
+    if not group.permissions.contains(permission):
+        group.permissions.add(permission)
+
+    # Modify the user's membership in the group based on the roles claim
+    is_user = settings.OIDC_OP_USER_ROLE in roles
+
+    if is_user and not user.groups.contains(group):
+        logger.debug(f"Adding {user.username} to {group.name}.")
+        user.groups.add(group)
+        user_modified = True
+    elif not is_user and user.groups.contains(group):
+        logger.debug(f"Removing {user.username} from {group.name}.")
+        user.groups.remove(group)
+        user_modified = True
+
+    return user_modified
+
+
+def process_roles_claim(user, roles):
+    """Adjust is_superuser and user group assignment based on OIDC roles"""
+    user_modified = process_roles_user(user, roles)
+    user_modified = process_roles_admin(user, roles) or user_modified
+    return user_modified
+
+
+def process_given_name_claim(user, given_name):
+    """Set a Django user's first_name to an OIDC given_name"""
+    if user.first_name != given_name:
+        user.first_name = given_name
+        return True
+
+
+def process_family_name_claim(user, family_name):
+    """Set a Django user's last_name to an OIDC family_name"""
+    if user.last_name != family_name:
+        user.last_name = family_name
+        return True
+
+
+class CFPBOIDCAuthenticationBackend(OIDCAuthenticationBackend):
+
+    def get_userinfo(self, access_token, id_token, payload):
+        # Roles are part of the parent payload here, and not the userinfo.
+        # Preserve them if we can.
+        roles = payload.get("roles", [])
+        userinfo = super().get_userinfo(access_token, id_token, payload)
+        userinfo["roles"] = roles
+        return userinfo
+
+    def process_claims(self, user, claims):
+        """Run a configured callable for each OIDC claim received
+
+        If a claim processor modifies a user it should return True, and the
+        user will be saved once all claims are processed."""
+        user_modified = False
+
+        claim_processors = getattr(settings, "OIDC_CLAIMS_PROCESSORS", {})
+        for claim, processor_string in claim_processors.items():
+            if claim not in claims:
+                continue
+
+            processor = import_string(processor_string)
+            processor_modified_user = processor(user, claims[claim])
+            user_modified = user_modified or processor_modified_user
```

Comment:
```suggestion
            user_modified = processor(user, claims[claim])
```

Reply:
With this change `user_modified` will only ever be the state of the past claims processor that touches the user, which could lose changes based on previous claims.

I was using a separate variable and a larger `for` loop for clarity's sake, but we could potentially use something like

```
processors = [(claim, import_string(processor_string)) for claim, processor_string in claim_processors.items() if claim in claims]
user_modified = any(processor(user, claims[claim]) for claim, processor in processors)
```

If we want more brevity here.

Reply:
Of course, you're right. How about

```suggestion
            user_modified |= processor(user, claims[claim])
```